File: sdks/python/src/ankka/agent.py
"""Agents: your process declares the instructions, the tools and the guardrails; the sidecar runs
the loop — the model call, tool dispatch, session memory, compaction and token accounting. Your
process is asked to plan an interaction, to run a tool, and to check a guardrail, and nothing
else: the model key lives on the sidecar, and this code never calls a model.
"""

from __future__ import annotations

import dataclasses
import inspect
import json
import types
import typing
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, ClassVar, get_args, get_origin, get_type_hints

from ankka._proto.ankka.protocol.v1 import discovery_pb2
from ankka.codec import Codec, default_codec_for
from ankka.context import Metadata
from ankka.effects.agent import AgentEffect, AgentEffects
from ankka.event_sourced_entity import HandlerSpec, RegistrationError, collect_handlers

if typing.TYPE_CHECKING:
    from ankka.client import ComponentClient

_STREAM = "_ankka_stream"


def stream(name: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
    """Declares a handler whose reply streams, token by token, under wire name ``name``."""

    def decorate(fn: Callable[..., Any]) -> Callable[..., Any]:
        setattr(fn, _STREAM, name)
        return fn

    return decorate


@dataclass(frozen=True)
class Tool:
    """A function the model may call. ``input`` is a dataclass (or ``None``) the model's arguments
    decode into — its fields are the schema the model sees — and ``run`` answers text for the
    model. An exception is a tool error, fed back to the model, which usually corrects itself."""

    description: str
    run: Callable[..., Any]
    input: Any = None

    def input_schema(self) -> str:
        return json.dumps(_schema_for(self.input) if self.input is not None else {"type": "object", "properties": {}})


@dataclass(frozen=True)
class Guardrail:
    """A check on what goes into, or comes out of, the model: ``check(stage, text)`` answers a
    reason to block, or ``None``. ``stage`` is ``"input"`` or ``"output"``."""

    check: Callable[[str, str], str | None]


def _schema_for(tp: Any) -> dict[str, Any]:
    origin = get_origin(tp)
    if origin in (types.UnionType, typing.Union):
        members = [a for a in get_args(tp) if a is not type(None)]
        if len(members) == 1:
            return _schema_for(members[0])
        return {"anyOf": [_schema_for(m) for m in members]}
    if tp is str:
        return {"type": "string"}
    if tp is bool:
        return {"type": "boolean"}
    if tp is int:
        return {"type": "integer"}
    if tp is float:
        return {"type": "number"}
    if origin in (list, tuple, set, frozenset):
        args = get_args(tp)
        return {"type": "array", "items": _schema_for(args[0]) if args else {}}
    if origin in (dict, typing.Mapping):
        args = get_args(tp)
        return {"type": "object", "additionalProperties": _schema_for(args[1]) if len(args) == 2 else {}}
    if dataclasses.is_dataclass(tp) and isinstance(tp, type):
        hints = get_type_hints(tp)
        properties = {f.name: _schema_for(hints.get(f.name, Any)) for f in dataclasses.fields(tp)}
        required = [
            f.name
            for f in dataclasses.fields(tp)
            if f.default is dataclasses.MISSING and f.default_factory is dataclasses.MISSING
        ]
        return {"type": "object", "properties": properties, "required": required, "additionalProperties": False}
    return {}


class Agent:
    """Subclass this: ``component_id``, ``tools = {name: Tool(...)}``, ``guardrails = {name:
    Guardrail(...)}``, and handlers decorated ``@command`` or ``@stream`` that return an
    ``AgentEffect`` built from ``self.effects``."""

    component_id: ClassVar[str]
    role: ClassVar[str | None] = None
    max_tool_call_steps: ClassVar[int] = 100
    tools: ClassVar[dict[str, Tool]] = {}
    guardrails: ClassVar[dict[str, Guardrail]] = {}
    _handlers: ClassVar[dict[str, HandlerSpec]]

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        if not hasattr(cls, "component_id"):
            raise RegistrationError(f"{cls.__name__} must declare component_id")
        handlers = collect_handlers(cls, effect_base=AgentEffect, read_only_base=AgentEffect)
        for attr, member in inspect.getmembers(cls, predicate=inspect.isfunction):
            name = getattr(member, _STREAM, None)
            if name is None:
                continue
            if name in handlers:
                raise RegistrationError(f"{cls.__name__}: handler '{name}' is declared twice")
            hints = get_type_hints(member)
            params = [p for p in inspect.signature(member).parameters.values() if p.name != "self"]
            if len(params) > 1:
                raise RegistrationError(f"{cls.__name__}.{attr}: a handler takes at most one input")
            input_type = hints.get(params[0].name) if params else None
            in_codec: Codec[Any] = default_codec_for(input_type) if input_type is not None else default_codec_for(type(None))
            handlers[name] = HandlerSpec(name, False, attr, input_type, str, in_codec, default_codec_for(str), streaming=True)
        cls._handlers = handlers
        for name, tool in cls.tools.items():
            if not tool.description:
                raise RegistrationError(f"{cls.__name__}: tool '{name}' needs a description; the model decides by it")
        if cls.max_tool_call_steps <= 0:
            raise RegistrationError(f"{cls.__name__}: max_tool_call_steps must be positive")

    def __init__(self, client: ComponentClient | None = None) -> None:
        self.effects = AgentEffects()
        self.client = client
        self._session_id = ""
        self._metadata = Metadata()

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def metadata(self) -> Metadata:
        return self._metadata

    @classmethod
    def handlers(cls) -> dict[str, HandlerSpec]:
        return cls._handlers

    @classmethod
    def to_component(cls) -> discovery_pb2.Component:
        return discovery_pb2.Component(
            kind=discovery_pb2.AGENT,
            id=cls.component_id,
            handlers=[h.to_pb() for h in sorted(cls._handlers.values(), key=lambda h: h.name)],
            agent=discovery_pb2.AgentDetail(
                role=cls.role or "",
                max_tool_call_steps=cls.max_tool_call_steps,
                tools=[
                    discovery_pb2.Tool(name=n, description=t.description, input_schema_json=t.input_schema())
                    for n, t in sorted(cls.tools.items())
                ],
                guardrails=sorted(cls.guardrails),
            ),
        )

    # ── What the server and the testkit drive ──────────────────────────────

    async def _plan(self, spec: HandlerSpec, input_bytes: bytes, session_id: str, metadata: Metadata) -> AgentEffect[Any]:
        self._session_id = session_id
        self._metadata = metadata
        method = getattr(self, spec.method_name)
        result = method() if spec.input_type is None else method(spec.input_codec.decode(input_bytes))
        if isinstance(result, Awaitable):
            result = await result
        if not isinstance(result, AgentEffect):
            raise TypeError(f"{type(self).__name__}.{spec.method_name} returned {type(result).__name__}, not an AgentEffect")
        effect = result
        unknown = [t for t in effect.tool_names if t not in type(self).tools] + [g for g in effect.guardrail_names if g not in type(self).guardrails]
        if unknown:
            raise RegistrationError(f"{type(self).__name__}: the plan names undeclared {', '.join(unknown)}")
        return effect

    async def _invoke_tool(self, name: str, arguments_json: str, session_id: str) -> str:
        self._session_id = session_id
        tool = type(self).tools[name]
        if tool.input is None:
            result = tool.run(self)
        else:
            result = tool.run(self, default_codec_for(tool.input).decode(arguments_json.encode("utf-8")))
        if isinstance(result, Awaitable):
            result = await result
        return result if isinstance(result, str) else json.dumps(result)

    async def _check_guardrail(self, name: str, stage: str, text: str, session_id: str) -> str | None:
        self._session_id = session_id
        result = type(self).guardrails[name].check(stage, text)
        if isinstance(result, Awaitable):
            result = await result
        return result

File: sdks/typescript/src/server/agent.ts
// The agent conversation: the sidecar asks this process to plan (an AgentEffect as data), to run a tool
// with the model's arguments, and to check a guardrail. The loop, the memory and the model's key are
// the sidecar's.

import type { ConnectRouter } from "@connectrpc/connect"
import { create } from "@bufbuild/protobuf"
import {
  Agent as AgentService,
  AgentPlan_Memory,
  GuardrailRequest_Stage,
  GuardrailResultSchema,
  PlanReplySchema,
  ToolResultSchema,
  type GuardrailRequest,
  type GuardrailResult,
  type PlanReply,
  type PlanRequest,
  type ToolRequest,
  type ToolResult,
} from "../_proto/ankka/protocol/v1/agent_pb.ts"
import { codecFor, isCodec, type Codec } from "../codec.ts"
import { metadataFromProto } from "../context.ts"
import { ErrorCode, type ErrorDetail } from "../effects/common.ts"
import type { AgentEffect } from "../effects/agent.ts"
import type { Agent } from "../agent.ts"
import { decodeJsonValue, reviver } from "../json.ts"
import { errorCodeToProto } from "../kinds.ts"
import type { Schema } from "../schema.ts"
import type { RegisteredAgent } from "../service.ts"
import { decodePayload } from "./payloads.ts"
import type { ServerContext } from "./server.ts"

function messageOf(e: unknown): string {
  return e instanceof Error ? e.message : String(e)
}

function planFailure(commandId: bigint, error: ErrorDetail): PlanReply {
  return create(PlanReplySchema, { message: { case: "failure", value: { commandId, error: { message: error.message, code: errorCodeToProto(error.code) } } } })
}

/** Checks a plan against the declaration: a plan may only name tools and guardrails discovery declared. */
export function checkPlan(registered: RegisteredAgent, effect: AgentEffect<unknown>): string | undefined {
  const unknown = [...effect.toolNames.filter((t) => !registered.tools.has(t)).map((t) => `tool ${JSON.stringify(t)}`), ...effect.guardrailNames.filter((g) => !registered.guardrails.has(g)).map((g) => `guardrail ${JSON.stringify(g)}`)]
  return unknown.length > 0 ? `${registered.id}: the plan names undeclared ${unknown.join(", ")}` : undefined
}

export async function handlePlan(req: PlanRequest, ctx: ServerContext): Promise<PlanReply> {
  const registered = ctx.registry.of("agent", req.componentId)
  if (!registered) return planFailure(0n, { message: `no agent ${JSON.stringify(req.componentId)} is registered`, code: ErrorCode.NotFound })
  const handler = registered.handlers.get(req.name)
  if (!handler) return planFailure(0n, { message: `no handler ${JSON.stringify(req.name)} on ${registered.id}`, code: ErrorCode.NotFound })
  const metadata = metadataFromProto(req.metadata)
  const agent = new registered.cls() as Agent
  agent._bind(req.sessionId, metadata, ctx.client.withMetadata(metadata))
  let effect: AgentEffect<unknown>
  try {
    const input = handler.input ? decodePayload(codecFor(handler.input), req.payload) : undefined
    effect = (await handler.run(agent, input)) as AgentEffect<unknown>
    if (effect?.kind !== "agent") throw new TypeError(`${registered.id}/${req.name} returned something that is not an agent effect`)
    const problem = checkPlan(registered, effect)
    if (problem) throw new Error(problem)
  } catch (e) {
    ctx.log(`ankka: agent ${registered.id}/${req.sessionId}/${req.name} threw: ${messageOf(e)}`)
    return planFailure(0n, { message: messageOf(e), code: ErrorCode.Internal })
  }
  return create(PlanReplySchema, {
    message: {
      case: "plan",
      value: {
        ...(effect.model !== null ? { model: effect.model } : {}),
        ...(effect.system !== null ? { system: effect.system } : {}),
        ...(effect.user !== null ? { user: effect.user } : {}),
        context: [...effect.context],
        memory: effect.sessionMemory ? AgentPlan_Memory.SESSION : AgentPlan_Memory.NONE,
        tools: [...effect.toolNames],
        guardrails: [...effect.guardrailNames],
        responseShape: effect.jsonReply ? { shape: { case: "json", value: { schemaHint: effect.schemaHint } } } : { shape: { case: "text", value: {} } },
        ...(effect.failure ? { failure: { message: effect.failure.message, code: errorCodeToProto(effect.failure.code) } } : {}),
      },
    },
  })
}

/** Runs a tool: the model's arguments, as JSON text, decoded with the tool's input shape; the result as text for the model. */
export async function runTool(registered: RegisteredAgent, agent: Agent, name: string, argumentsJson: string): Promise<string> {
  const tool = registered.tools.get(name)
  if (!tool) throw new Error(`no tool ${JSON.stringify(name)} on ${registered.id}`)
  let input: unknown
  if (isCodec(tool.input)) input = (tool.input as Codec<unknown>).decode(new TextEncoder().encode(argumentsJson))
  else input = decodeJsonValue(tool.input as Schema<unknown>, JSON.parse(argumentsJson || "{}", reviver))
  const result = await tool.run(agent, input)
  return typeof result === "string" ? result : JSON.stringify(result, (_k, v) => (typeof v === "bigint" ? v.toString() : v))
}

export async function handleTool(req: ToolRequest, ctx: ServerContext): Promise<ToolResult> {
  const registered = ctx.registry.of("agent", req.componentId)
  if (!registered) return create(ToolResultSchema, { result: { case: "error", value: `no agent ${JSON.stringify(req.componentId)} is registered` } })
  const agent = new registered.cls() as Agent
  agent._bind(req.sessionId, {}, ctx.client)
  try {
    return create(ToolResultSchema, { result: { case: "ok", value: await runTool(registered, agent, req.tool, req.argumentsJson) } })
  } catch (e) {
    // A tool error is fed back to the model, which usually corrects itself.
    return create(ToolResultSchema, { result: { case: "error", value: messageOf(e) } })
  }
}

export async function handleGuardrail(req: GuardrailRequest, ctx: ServerContext): Promise<GuardrailResult> {
  const registered = ctx.registry.of("agent", req.componentId)
  const guardrail = registered?.guardrails.get(req.guardrail)
  if (!registered || !guardrail) return create(GuardrailResultSchema, { result: { case: "block", value: `no guardrail ${JSON.stringify(req.guardrail)} on ${req.componentId}` } })
  try {
    const reason = await guardrail.check(req.stage === GuardrailRequest_Stage.OUTPUT ? "output" : "input", req.text)
    return reason === null || reason === undefined
      ? create(GuardrailResultSchema, { result: { case: "pass", value: {} } })
      : create(GuardrailResultSchema, { result: { case: "block", value: reason } })
  } catch (e) {
    ctx.log(`ankka: guardrail ${registered.id}/${req.guardrail} threw: ${messageOf(e)}`)
    return create(GuardrailResultSchema, { result: { case: "block", value: `guardrail ${req.guardrail} failed: ${messageOf(e)}` } })
  }
}

export function agentRoutes(router: ConnectRouter, ctx: ServerContext): void {
  router.service(AgentService, {
    plan: (req) => handlePlan(req, ctx),
    invokeTool: (req) => handleTool(req, ctx),
    checkGuardrail: (req) => handleGuardrail(req, ctx),
  })
}
